packages/@react-aria/dnd/src/useDroppableCollection.ts

```diff
@@ -0,0 +1,396 @@
+      if (!target) {
+        return {
+          type: 'root'
+        };
+      }
+
+      let {keyboardDelegate} = localState.props;
+      let nextKey = target.type === 'item'
+        ? keyboardDelegate.getKeyBelow(target.key)
+        : keyboardDelegate.getFirstKey();
+
+      if (target.type === 'item') {
+        let positionIndex = DROP_POSITIONS.indexOf(target.dropPosition);
+        let nextDropPosition = DROP_POSITIONS[positionIndex + 1];
+        if (positionIndex < DROP_POSITIONS.length - 1 && !(nextDropPosition === 'after' && nextKey != null)) {
+          return {
+            type: 'item',
+            key: target.key,
+            dropPosition: nextDropPosition
+          };
+        }
+      }
+
+      if (nextKey == null) {
+        if (wrap) {
+          return {
+            type: 'root'
+          };
+        }
+
+        return null;
+      }
+
+      return {
+        type: 'item',
+        key: nextKey,
+        dropPosition: DROP_POSITIONS[0]
+      };
+    };
+
+    let getPreviousTarget = (target: DropTarget, wrap = true): DropTarget => {
+      let {keyboardDelegate} = localState.props;
+      let nextKey = target?.type === 'item'
+        ? keyboardDelegate.getKeyAbove(target.key)
+        : keyboardDelegate.getLastKey();
+
+      if (target?.type === 'item') {
+        let positionIndex = DROP_POSITIONS.indexOf(target.dropPosition);
+        let nextDropPosition = DROP_POSITIONS[positionIndex - 1];
+        if (positionIndex > 0 && nextDropPosition !== 'after') {
+          return {
+            type: 'item',
+            key: target.key,
+            dropPosition: nextDropPosition
+          };
+        }
+      }
+
+      if (nextKey == null) {
+        if (wrap) {
+          return {
+            type: 'root'
+          };
+        }
+
+        return null;
+      }
+
+      return {
+        type: 'item',
+        key: nextKey,
+        dropPosition: !target || target.type === 'root' ? 'after' : 'on'
+      };
+    };
+
+    let nextValidTarget = (
+      target: DropTarget,
+      types: Set<string>,
+      allowedDropOperations: DropOperation[],
+      getNextTarget: (target: DropTarget, wrap: boolean) => DropTarget,
+      wrap = true
+    ): DropTarget => {
+      let seenRoot = 0;
+      let operation: DropOperation;
+      do {
+        let nextTarget = getNextTarget(target, wrap);
+        if (!nextTarget) {
+          return null;
+        }
+        target = nextTarget;
+        operation = localState.state.getDropOperation(nextTarget, types, allowedDropOperations);
+        if (target.type === 'root') {
+          seenRoot++;
+        }
+      } while (
+        target &&
+        operation === 'cancel' &&
+        !localState.state.isDropTarget(target) &&
+        seenRoot < 2
+      );
+
+      if (operation === 'cancel') {
+        return null;
+      }
+
+      return target;
+    };
+
+    DragManager.registerDropTarget({
```

Comment:
i think the cleanup part of this is never called?

Reply:
Good catch